plugins/journald/fu-plugin-journald.c

```diff
@@ -0,0 +1,91 @@
+/*
+ * Copyright (C) 2020 Richard Hughes <richard@hughsie.com>
+ *
+ * SPDX-License-Identifier: LGPL-2.1+
+ */
+
+#include "config.h"
+
+#include <string.h>
+#include <systemd/sd-journal.h>
+
+#include "fu-plugin-vfuncs.h"
+#include "fu-hash.h"
+
+struct FuPluginData {
+	gboolean		 has_tme;
+};
+
+void
+fu_plugin_init (FuPlugin *plugin)
+{
+	fu_plugin_alloc_data (plugin, sizeof (FuPluginData));
+	fu_plugin_set_build_hash (plugin, FU_BUILD_HASH);
+}
+
+G_DEFINE_AUTOPTR_CLEANUP_FUNC(sd_journal, sd_journal_close)
+
+gboolean
+fu_plugin_startup (FuPlugin *plugin, GError **error)
+{
+	FuPluginData *priv = fu_plugin_get_data (plugin);
+	const void *buf = NULL;
+	gint rc;
+	gsize bufsz = 0;
+	g_autoptr(sd_journal) j = NULL;
+
+	/* read entire journal */
+	rc = sd_journal_open (&j, SD_JOURNAL_SYSTEM);
+	if (rc < 0) {
+		g_set_error (error,
+			     FWUPD_ERROR,
+			     FWUPD_ERROR_NOT_SUPPORTED,
+			     "failed to open journal: %s [%i]",
+			     strerror (rc), rc);
+		return FALSE;
+	}
+	while (sd_journal_next (j) > 0) {
```

Comment:
I'm a little worried this turns into a performance bottleneck.  Is there a way to filter it to only be running from the current boot?  Or only from time the kernel finished initializing?

Reply:
It takes 68ms on my Lenovo P50, which is a pretty powerful machine. I can imagine something like a raspberry pi might take a lot longer.

Reply:
Of course, we could just read the journal when getting the hsi report... will fix.

Reply:
Well I think when paired with `sd_id128_get_boot`, `sd_journal_seek_monotonic_usec` it's possible to make it a super light operation for most people even with a large journal.  It would at least allow only looking at the current boot's journal.

Reply:
Hmm, I tried that and didn't get any speedup at all.